cpp/src/Ice/ObjectAdapterFactory.cpp

```diff
@@ -25,16 +26,13 @@ IceInternal::ObjectAdapterFactory::shutdown()
         // Ignore shutdown requests if the object adapter factory has
         // already been shut down.
         //
-        if (!_instance)
+        if (_isShutdown)
         {
             return;
         }
+        _isShutdown = true;
 
         adapters = _adapters;
-
-        _instance = nullptr;
```

Comment:
Refactored this code to avoid calling `instance->getAdmin()` with `_mutex` locked. (even putting aside lock-acquisition order concerns, getAdmin should be async and therefore should not be called within a mutex lock).